presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/RewriteAggregationIfToFilter.java

```diff
@@ -0,0 +1,197 @@
+            VariableReferenceExpression outputVariable = entry.getKey();
+            SpecialFormExpression ifExpression = (SpecialFormExpression) entry.getValue();
+
+            RowExpression condition = ifExpression.getArguments().get(0);
+            VariableReferenceExpression conditionReference = context.getVariableAllocator().newVariable(condition);
+            newAssignments.put(conditionReference, condition);
+            aggregationReferenceToConditionReference.put(outputVariable, conditionReference);
+
+            RowExpression trueResult = ifExpression.getArguments().get(1);
+            VariableReferenceExpression ifResultReference = context.getVariableAllocator().newVariable(trueResult);
+            newAssignments.put(ifResultReference, trueResult);
+            aggregationReferenceToIfResultReference.put(outputVariable, ifResultReference);
+        }
+
+        // Build new aggregations.
+        ImmutableMap.Builder<VariableReferenceExpression, Aggregation> aggregations = ImmutableMap.builder();
+        // Stores the masks used to build the filter predicates. Use set to dedup the predicates.
+        ImmutableSortedSet.Builder<VariableReferenceExpression> masks = ImmutableSortedSet.naturalOrder();
+        for (Map.Entry<VariableReferenceExpression, Aggregation> entry : aggregationNode.getAggregations().entrySet()) {
+            VariableReferenceExpression output = entry.getKey();
+            Aggregation aggregation = entry.getValue();
+            if (!aggregationsToRewrite.contains(aggregation)) {
+                aggregations.put(output, aggregation);
+                continue;
+            }
+            VariableReferenceExpression aggregationReference = (VariableReferenceExpression) aggregation.getArguments().get(0);
+            CallExpression callExpression = aggregation.getCall();
+            VariableReferenceExpression mask = aggregationReferenceToConditionReference.get(aggregationReference);
+            aggregations.put(output, new Aggregation(
+                    new CallExpression(
+                            callExpression.getDisplayName(),
+                            callExpression.getFunctionHandle(),
+                            callExpression.getType(),
+                            ImmutableList.of(aggregationReferenceToIfResultReference.get(aggregationReference))),
+                    Optional.empty(),
+                    aggregation.getOrderBy(),
+                    aggregation.isDistinct(),
+                    Optional.of(aggregationReferenceToConditionReference.get(aggregationReference))));
+            masks.add(mask);
+        }
+
+        RowExpression predicate = TRUE_CONSTANT;
+        if (!aggregationNode.hasNonEmptyGroupingSet() && aggregationsToRewrite.size() == aggregationNode.getAggregations().size()) {
+            // All aggregations are rewritten by this rule. We can add a filter with all the masks to make the query more efficient.
+            predicate = or(masks.build());
```

Comment:
This or might actually cause more slowdown than help because all masks have to be evaluated for every row anyway. I don't think this helps.

Reply:
I think it helps because
* The filter could be pushed down to the scan node which might be able to evaluated very efficiently, e.g., if it is on some partition columns, the partition columns are using dictionary encoding. This only need to evaluated once per dictionary item. 
* The filter can be used to prune the partitions/splits based on column stats.

Besides, the AGG() FILTER implementation also adds this predicate. It is better to keep the same behavior:
https://github.com/prestodb/presto/blob/04138bc26a4fc801630fe936463d89d071d00c0a/presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/ImplementFilteredAggregations.java#L121